src/msft-recommendation/src/main/scala/TrainValidRecommendSplit.scala

```diff
@@ -0,0 +1,245 @@
+    .drop("nitems")
+
+  private def filterByUserRatingCount(dataset: Dataset[_]): DataFrame = dataset
+    .groupBy($(itemCol))
+    .agg(col($(itemCol)), count(col($(userCol))))
+    .withColumnRenamed("count(" + $(userCol) + ")", "ncustomers")
+    .where(col("ncustomers") >= $(minRatingsI))
+    .join(dataset, $(itemCol))
+    .drop("ncustomers")
+
+  private def filterRatings(dataset: Dataset[_]): DataFrame = filterByUserRatingCount(dataset)
+    .join(filterByItemCount(dataset), $(userCol))
+
+  def splitDF(dataset: DataFrame): Array[DataFrame] = {
+    import dataset.sqlContext.implicits._
+
+    val (tr_idx: RDD[(Any, List[Any])], testIndex: RDD[(Any, List[Any])]) = {
+      def permIndicesLambda(r: Row): (Any, List[Any], List[Any]) =
+        (r(0), Random.shuffle(List(r(1))), List(r(1)))
+
+      def firstN(r: (Any, List[Any], List[Any])): (Any, List[Any]) =
+        (r._1, r._2.slice(0, math.round(r._3.size.toDouble * $(trainRatio)).toInt))
+
+      def lastN(r: (Any, List[Any], List[Any])): (Any, List[Any]) =
+        (r._1, r._2.drop(math.round(r._3.size.toDouble * $(trainRatio)).toInt))
+
+      val perm_indices: RDD[(Any, List[Any], List[Any])] = dataset
+        .groupBy($(itemCol))
+        .agg(col($(itemCol)), count($(userCol)))
+        .withColumnRenamed("count(" + $(userCol) + ")", "nusers").rdd
+        .map(permIndicesLambda)
+        .cache()
+
+      val tr_idx: RDD[(Any, List[Any])] = perm_indices.map(firstN)
+      val testIndex: RDD[(Any, List[Any])] = perm_indices.map(lastN)
+      perm_indices.unpersist()
+
+      (tr_idx, testIndex)
+    }
+
+    val (trainingDataset: DataFrame, validationDataset: DataFrame) = {
+      def validationFlaptMapLambda(r: (Any, (Iterable[Row], List[Any]))): Iterable[Row] =
+        r._2._1.drop(r._2._2.size)
+
+      def mapLambda(r: Row): (Int, Int, String, String, Int) =
+        (r.getDouble(0).toInt, r.getDouble(1).toInt, r.getString(2), r.getString(3), r.getInt(4))
+
+      def flatMapLambda(r: (Any, (Iterable[Row], List[Any]))): Iterable[Row] =
+        r._2._1.slice(0, r._2._2.size)
+
+      val trainingDataset = dataset.rdd
+        .groupBy(_ (0))
+        .join(tr_idx)
+        .flatMap(flatMapLambda)
+        .map(mapLambda)
+        .toDF($(userCol), $(itemCol), $(userCol) + "Org", $(itemCol) + "Org", $(ratingCol))
+
+      val validationDataset: DataFrame = dataset.rdd
+        .groupBy(_ (1))
+        .join(testIndex)
+        .flatMap(validationFlaptMapLambda)
+        .map(mapLambda)
+        .toDF($(userCol), $(itemCol), $(userCol) + "Org", $(itemCol) + "Org", $(ratingCol))
+
+      (trainingDataset, validationDataset)
+    }
+
+    Array(trainingDataset, validationDataset)
+  }
+
+  def prepareTestData(validationDataset: DataFrame, recs: DataFrame): Dataset[_] = {
+    import org.apache.spark.sql.functions.{collect_list, rank => r}
+
+    val est = $(estimator) match {
+      case p: Pipeline => {
+        //Assume Rec is last stage of pipeline
```

Comment:
This is a somewhat unusual requirement - what is the use-case for passing in a pipeline as the estimator here?